File: Python/Backtracking/num_islands_I.py
# Counting the number of islands using Depth First Search and Backtracking
# Author: Pavan Kumar Paluri
# LeetCode - Medium

class Solution:
    def dfs_helper(self, grid, i, j):
        # Check for boundary conditions
        if i<0 or i>=len(grid) or j<0 or j>=len(grid[i]) or grid[i][j]=="0":
            return 0
        # once here, we only face 1:
        # if that is visited, sink the island-> as in make it 0.
        grid[i][j]="0"
        # now explore in all four directions
        self.dfs_helper(grid, i+1, j) # south
        self.dfs_helper(grid, i-1, j) # north
        self.dfs_helper(grid, i, j+1) # left 
        self.dfs_helper(grid, i, j-1) # right
        return 1
    
    def numIslands(self, grid: List[List[str]]) -> int:
        # check for base condition
        if grid is None or len(grid) ==0:
            # basically no islands exist
            return 0
        # now traverse through the list to add numof islands
        numIslands =0
        for i in range(0, len(grid)): # rows
            for j in range(0, len(grid[i])): # cols
                numIslands += self.dfs_helper(grid, i, j)
        
        return numIslands

File: C++/Backtracking/num_islands_I.cpp
// Counting the number of islands in a given 2D matrix
// Author: Pavan Kumar Paluri
// Leetcode - Medium
// Approach used: DFS + Backtracking

class Solution {
public:
    int dfs_helper(vector<vector<char>>&grid, int row, int col)
    {
        // chack for base conditions
        if(row<0 || row>=grid.size() || col<0 || col>=grid[row].size() || grid[row][col]=='0') 
            return 0;
        // if here, valid cases only - as in islands with 1
        // if an island is visited- sink it- change it to 0
        grid[row][col]='0';
        // explore cardinal directions to see if we can get extended islands
        dfs_helper(grid, row+1, col); // down
        dfs_helper(grid, row-1, col); // up
        dfs_helper(grid, row, col+1); // right
        dfs_helper(grid, row, col-1); // left 
        return 1;
    }
    
    int numIslands(vector<vector<char>>& grid) {
        // check base conditions
        if(grid.size() == 0 || grid.empty())
            return 0;
        // if here: grid is valid one
        int num_islands  =0;
        for(int i=0; i<grid.size(); i++)
        {
            for(int j=0; j<grid[i].size(); j++)
            {
                num_islands += dfs_helper(grid, i, j);
            }
        }
        
        return num_islands;
    }
};
